homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,288 @@
+
+        if enasolar.sun_up <= datetime.now().time() <= enasolar.sun_down:
+            values = await enasolar.read_data()
+        else:
+            values = False
+
+        for sensor in hass_data_sensors:
+            state_unknown = False
+            if not values and (
+                (sensor.per_day_basis and date.today() > sensor.date_updated)
+                or (not sensor.per_day_basis and not sensor.per_total_basis)
+            ):
+                state_unknown = True
+            sensor.async_update_values(unknown_state=state_unknown)
+            _LOGGER.debug(
+                "Data Sensor %s updated => %s", sensor._sensor.key, sensor.state
+            )
+
+        return values
+
+    def start_update_interval(event):
+        """Start the update interval scheduling."""
+        nonlocal remove_interval_update_meters, remove_interval_update_data
+
+        remove_interval_update_meters = async_track_time_interval_backoff(
+            hass, async_enasolar_meters, SCAN_METERS_MIN_INTERVAL
+        )
+        remove_interval_update_data = async_track_time_interval_backoff(
+            hass, async_enasolar_data, SCAN_DATA_MIN_INTERVAL
+        )
+
+    def stop_update_interval(event):
+        """Properly cancel the scheduled update."""
+        remove_interval_update_meters()  # pylint: disable=not-callable
+        remove_interval_update_data()  # pylint: disable=not-callable
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, stop_update_interval)
+
+
+@callback
+def async_track_time_interval_backoff(hass, action, min_interval) -> CALLBACK_TYPE:
+    """Add a listener that fires repetitively and increases the interval when failed."""
+    remove = None
+    interval = min_interval
+
+    async def interval_listener(now=None):
+        """Handle elapsed interval with backoff."""
+        nonlocal interval, remove
+        try:
+            if await action():
+                interval = min_interval
+            else:
+                interval = min(interval * 2, SCAN_MAX_INTERVAL)
+        finally:
+            remove = async_call_later(hass, interval, interval_listener)
+
+    hass.async_create_task(interval_listener())
+
+    def remove_listener():
+        """Remove interval listener."""
+        if remove:
+            remove()  # pylint: disable=not-callable
+
+    return remove_listener
+
+
+class EnaSolarSensor(SensorEntity):
+    """Representation of a EnaSolar sensor."""
+
+    def __init__(self, pyenasolar_sensor, inverter_name=None):
+        """Initialize the EnaSolar sensor."""
+        self._sensor = pyenasolar_sensor
+        self._inverter_name = inverter_name
+        self._state = self._sensor.value
+
+        if pyenasolar_sensor.is_meter:
+            self._attr_state_class = STATE_CLASS_MEASUREMENT
+        else:
+            self._attr_state_class = STATE_CLASS_TOTAL_INCREASING
+
+    @property
+    def name(self):
```

Comment:
If the name doesn't change, you can simply set `self._attr_name` in `__init__`

Reply:
It docent change though is optional, I might leave it like it is for now